[Driver][SYCL] Add initial SYCL offload compilation support

mdtoguchi opened the pull request:
Introduces the SYCL based toolchain and initial toolchain construction when using the '-fsycl' option.  This option will enable SYCL based offloading, creating a SPIR-V based IR file packaged into the compiled host object.

This includes early support for creating the host/device object using the new offloading model.  The device object is created using the spir64-unknown-unknown target triple.

New/Updated Options:
 -fsycl  Enables SYCL offloading for host and device
 -fsycl-device-only
         Enables device only compilation for SYCL
 -fsycl-host-only
         Enables host only compilation for SYCL

RFC Reference: https://discourse.llvm.org/t/rfc-sycl-driver-enhancements/74092

mdtoguchi committed:
[Driver][SYCL] Add initial SYCL offload compilation support

Introduces the SYCL based toolchain and initial toolchain construction
when using the '-fsycl' option.  This option will enable SYCL based
offloading, creating a SPIR-V based IR file packaged into the compiled
host object.

This includes early support for creating the host/device object using
the new offloading model.  The device object is created using the
spir64-unknown-unknown target triple.

New/Updated Options:
 -fsycl  Enables SYCL offloading for host and device
 -fsycl-device-only
         Enables device only compilation for SYCL
 -fsycl-host-only
         Enables host only compilation for SYCL

RFC Reference: https://discourse.llvm.org/t/rfc-sycl-driver-enhancements/74092

asudarsa commented on clang/include/clang/Driver/Action.h:
```diff
@@ -94,6 +94,7 @@ class Action {
     OFK_Cuda = 0x02,
     OFK_OpenMP = 0x04,
     OFK_HIP = 0x08,
+    OFK_SYCL = 0x10
```
Minor nit:
```suggestion
    OFK_SYCL = 0x10,
```

mdtoguchi replied:
Updated with latest commit

asudarsa commented on clang/include/clang/Basic/DiagnosticDriverKinds.td:
```diff
@@ -195,6 +195,8 @@ def err_drv_Xopenmp_target_missing_triple : Error<
   "cannot deduce implicit triple value for -Xopenmp-target, specify triple using -Xopenmp-target=<triple>">;
 def err_drv_invalid_Xopenmp_target_with_args : Error<
   "invalid -Xopenmp-target argument: '%0', options requiring arguments are unsupported">;
+def err_drv_unsupported_opt_sycl : Error<
```
Do we really need a SYCL specific error here? Can we piggyback on existing error format (err_drv_argument_only_allowed_with)?

Thanks

mdtoguchi replied:
That's a good point.  There is an inverse `err_drv_argument_not_allowed_with` as well.  I will see about updating this.

mdtoguchi committed:
Address review comments

use a more common diagnostic
Update tests to check diagnostic and missed header check

mdtoguchi committed:
Cleanup unused option

asudarsa commented on clang/include/clang/Driver/Driver.h:
```diff
@@ -582,6 +582,10 @@ class Driver {
 
+  /// MakeSYCLDeviceTriple - Returns the SYCL device triple for the
+  /// specified subarch
+  llvm::Triple MakeSYCLDeviceTriple(StringRef TargetArch = "spir64") const;
```
Will it be possible to use spirv64 as default here? Thanks

mdtoguchi replied:
Using `spirv64` wasn't discussed as a default.  @bader do you have any thoughts on this?

bader replied:
I think using spirv64 by default in the long term makes sense. `spir64` is associated with the LLVM bitcode format following the SPIR-1.2 spec (i.e. using spir/spi64 target triple, certain metadata, etc.) adjusted to recent LLVM IR version.

@asudarsa, do you imply that we should use SPIR-V back-end to produce the SPIR-V instead of spir-targeted LLVM IR?

asudarsa replied:
Hi @bader 

Yes. That's the eventual goal. However, we are not yet ready to transition from the SPIR-V Translator to the Backend. I think we should be able to use spirv64 with the translator as well. 

Thanks

mdtoguchi replied:
Offline quick convo with @asudarsa concluded we will stick with spir64 initially.

asudarsa commented on clang/include/clang/Driver/Options.td:
```diff
@@ -6691,12 +6692,16 @@ defm : FlangIgnoredDiagOpt<"frontend-loop-interchange">;
 defm : FlangIgnoredDiagOpt<"target-lifetime">;
 
 // C++ SYCL options
+let Group = sycl_Group in {
```
Please take a look at few more SYCL options listed here and check if we need to clean that up. Thanks
https://github.com/mdtoguchi/llvm-project/blob/d819f8911ad6360b667289bd90eb5b6401105335/clang/include/clang/Driver/Options.td#L8204

mdtoguchi replied:
This section is for `-cc1` options.  The only potential cleanup is the values for `-sycl-std` but we can do that in a future patch.

asudarsa commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -767,6 +768,26 @@ Driver::OpenMPRuntimeKind Driver::getOpenMPRuntime(const ArgList &Args) const {
   return RT;
 }
 
+static const char *getDefaultSYCLArch(Compilation &C) {
```
Same question as above about default arch triple. Thanks

srividya-sundaram replied:
nit:

asudarsa commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -979,6 +1000,45 @@ void Driver::CreateOffloadingDeviceToolChains(Compilation &C,
+  //
+  // SYCL
+  //
+  // We need to generate a SYCL toolchain if the user specified -fsycl.
```
I think this PR will benefit from updating SYCLSupport.rst document. I can help with that if need be. Thanks

asudarsa commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -979,6 +1000,45 @@ void Driver::CreateOffloadingDeviceToolChains(Compilation &C,
+    addSYCLDefaultTriple(C, UniqueSYCLTriplesVec);
+  }
+
+  // We'll need to use the SYCL and host triples as the key into
```
Do we need to guard this somehow so that it gets called for SYCL compilation? I see IsCuda and IsHIP being computed for those offloading models. Please take a look at this check and see if we need something similar.
https://github.com/mdtoguchi/llvm-project/blob/d819f8911ad6360b667289bd90eb5b6401105335/clang/lib/Driver/Driver.cpp#L812

mdtoguchi replied:
I can add a guard.

asudarsa commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -6553,6 +6635,19 @@ const ToolChain &Driver::getOffloadingDeviceToolChain(
                                                            HostTC, Args);
       break;
     }
+    case Action::OFK_SYCL:
```
May be a TODO here that additional targets will be added?

mdtoguchi replied:
Addressed in recent commit.

asudarsa commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -1073,6 +1074,8 @@ void Clang::AddPreprocessingOptions(Compilation &C, const JobAction &JA,
   if (JA.isOffloading(Action::OFK_HIP))
     getToolChain().AddHIPIncludeArgs(Args, CmdArgs);
+  if (JA.isOffloading(Action::OFK_SYCL))
+    toolchains::SYCLToolChain::AddSYCLIncludeArgs(D, Args, CmdArgs);
```
Why can we not do a 'getToolChain() for SYCL? Thanks

mdtoguchi replied:
We can probably match up here, the use of getToolChain() is dependent on the given toolchain so it should be fine as we are in the SYCLToolChain here.

mdtoguchi replied:
I have made some significant updates here to sync things up a bit better to allow for `getToolChain()` usage.

asudarsa commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -5116,15 +5123,39 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
+  Arg *SYCLStdArg = Args.getLastArg(options::OPT_sycl_std_EQ);
+
+  if (IsSYCLOffloadDevice) {
+    // Pass the triple of host when doing SYCL
```
Please try repharasing this comment. Sounds a bit vague. Thanks

asudarsa commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -5116,15 +5123,39 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
+    // Add any options that are needed specific to SYCL offload while
+    // performing the host side compilation.
+    if (!IsSYCLOffloadDevice) {
+      // Let the FE know we are doing a SYCL offload compilation, but we are
+      // doing the host pass.
```
```suggestion
      // Let the front-end host compilation flow know about SYCL offload compilation.
```

asudarsa commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -6061,7 +6092,7 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
   // `--gpu-use-aux-triple-only` is specified.
   if (!Args.getLastArg(options::OPT_gpu_use_aux_triple_only) &&
-      (IsCudaDevice || IsHIPDevice)) {
+      (IsCudaDevice || IsHIPDevice || IsSYCLOffloadDevice)) {
```
Nit:
```suggestion
      (IsCudaDevice || IsHIPDevice || IsSYCLDevice)) {
```

mdtoguchi commented:
@bader - if you have any additional comments, they are welcome.  I plan on making pushing this to the main llvm-project once comments are addressed.

srividya-sundaram commented on clang/include/clang/Driver/Driver.h:
```diff
@@ -582,6 +582,10 @@ class Driver {
   /// @{
 
+  /// MakeSYCLDeviceTriple - Returns the SYCL device triple for the
+  /// specified subarch
```
nit: We are passing ArchType not subarch.

srividya-sundaram commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -979,6 +1000,45 @@ void Driver::CreateOffloadingDeviceToolChains(Compilation &C,
+  // SYCL
+  //
+  // We need to generate a SYCL toolchain if the user specified -fsycl.
+  // If -fsycl is supplied without any of these we will assume SPIR-V.
```
nit: not clear what 'any of these' refers to.
Could be re-phrased to something like: `If -fsycl-targets is not specified, then single SPIR-V target is assumed`

mdtoguchi replied:
This initial PR will not introduce `-fsycl-targets`, we will just rely on `-fsycl` to imply `spir64` for JIT.  I will clean up the comments.

bader commented on clang/lib/Driver/ToolChains/SYCL.h:
```diff
@@ -0,0 +1,68 @@
+//===--- SYCL.h - SYCL ToolChain Implementations -----------------*- C++ -*-===//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
```
```suggestion
//===--- SYCL.h - SYCL ToolChain Implementations -----------------*- C++ -*-===//
//
// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
```

srividya-sundaram commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -979,6 +1000,45 @@ void Driver::CreateOffloadingDeviceToolChains(Compilation &C,
+  // getOffloadingDeviceToolChain, because the device toolchains we're
+  // going to create will depend on both.
+  const ToolChain *HostTC = C.getSingleOffloadToolChain<Action::OFK_Host>();
+  for (auto &TT : UniqueSYCLTriplesVec) {
```
should this be `'const' `?

bader commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -979,6 +1000,45 @@ void Driver::CreateOffloadingDeviceToolChains(Compilation &C,
+  //
+  // We need to generate a SYCL toolchain if the user specified -fsycl.
+  // If -fsycl is supplied without any of these we will assume SPIR-V.
+  bool HasValidSYCLRuntime =
+      C.getInputArgs().hasFlag(options::OPT_fsycl, options::OPT_fno_sycl,
+                               false);
+
+  auto argSYCLIncompatible = [&](OptSpecifier OptId) {
+    if (!HasValidSYCLRuntime)
+      return;
+    if (Arg *IncompatArg = C.getInputArgs().getLastArg(OptId))
+      Diag(clang::diag::err_drv_argument_not_allowed_with)
+          << IncompatArg->getSpelling() << "-fsycl";
+  };
+  // -static-libstdc++ is not compatible with -fsycl.
+  argSYCLIncompatible(options::OPT_static_libstdcxx);
+  // -ffreestanding cannot be used with -fsycl
+  argSYCLIncompatible(options::OPT_ffreestanding);
```
Using `HasValidSYCLRuntime` name for the `hasFlag(OPT_fsycl)` return value is fine only for these few lines. These are checking the compatibility with the flags which configure different runtime settings.

But later code looks a bit strange. We probably should use something like `IsSYCL` as HIP/CUDA does here: https://github.com/mdtoguchi/llvm-project/blob/d819f8911ad6360b667289bd90eb5b6401105335/clang/lib/Driver/Driver.cpp#L799-L809.

mdtoguchi replied:
Updated to use `IsSYCL`

bader reviewed (commented):
@mdtoguchi, thanks!
No major concerns from my side. Let's open for review by the community.

mdtoguchi committed:
Adjust header inclusion and update variable name

mdtoguchi committed:
Address a few review comments

asudarsa commented on llvm/include/llvm/TargetParser/Triple.h:
```diff
@@ -831,6 +831,8 @@ class Triple {
            getArch() == Triple::spirv;
   }
 
+  bool isSPIROrSPIRV() const { return isSPIR() || isSPIRV(); }
```
This can benefit from a comment (either a code comment or PR comment)
// Currently, we use spir-based target triples to represent JIT compilation targets for SYCL. We will transition to using spir-v based target triples to represent JIT compilation targets for SYCL very soon. This helper function is used (instead of isSPIR) to ease that transition.

asudarsa commented on clang/lib/Driver/ToolChains/SYCL.cpp:
```diff
@@ -0,0 +1,202 @@
+void SYCLInstallationDetector::AddSYCLIncludeArgs(
+    const ArgList &DriverArgs, ArgStringList &CC1Args) const {
+  // Add the SYCL header search locations in the specified order.
+  //   ../include/sycl
```
Instead of relative path, can we specify some absolute path? In other implementations, I see something like INSTALL_PATH. Thanks

mdtoguchi replied:
This does resolve to an absolute path (D.Dir() is a full path) so the information isn't relative to any binary.  Current implementation is finding the headers in `../include` so if we end up needing to resolve in a different location that is build-time defined we can do so.

asudarsa commented on clang/lib/Driver/ToolChains/SYCL.cpp:
```diff
@@ -0,0 +1,202 @@
+  }
+}
+
+// Unsupported options for device compilation
```
Though this is SYCL specific file, I am thinking if it might be worthwhile to say 'SYCL device compilation' in comment here. A nit. Thanks

asudarsa commented on clang/lib/Driver/ToolChains/SYCL.cpp:
```diff
@@ -0,0 +1,202 @@
+SYCLToolChain::SYCLToolChain(const Driver &D, const llvm::Triple &Triple,
+                             const ToolChain &HostTC, const ArgList &Args)
+    : ToolChain(D, Triple, Args), HostTC(HostTC), SYCLInstallation(D) {
+  // Lookup binaries into the driver directory, this is used to
```
Do we need this? We are not going to be using clang-offload-bundler in SYCL compilation flow.

mdtoguchi replied:
We do use the `clang-offload-packager` though, which should also reside here.  I'll clean up the comment to be more generic for 'supporting tools'

asudarsa commented on clang/lib/Driver/ToolChains/MSVC.cpp:
```diff
@@ -427,7 +427,7 @@ void visualstudio::Linker::ConstructJob(Compilation &C, const JobAction &JA,
                              const ArgList &Args)
     : ToolChain(D, Triple, Args), CudaInstallation(D, Triple, Args),
-      RocmInstallation(D, Triple, Args) {
+      RocmInstallation(D, Triple, Args), SYCLInstallation(D) {
```
Why do we not pass Triple and Args for SYCL Installation?

mdtoguchi replied:
The `Triple` and `Args` are not used with our installation setup, so there is no need to add them.

asudarsa commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -1023,20 +1023,19 @@ void Driver::CreateOffloadingDeviceToolChains(Compilation &C,
-  // If -fsycl is supplied without -fsycl-targets we will assume SPIR-V.
-  // For -fsycl-device-only, we also setup the implied triple as needed.
-  if (HasValidSYCLRuntime) {
+  // If -fsycl is supplied we will assume SPIR-V.
```
I am a bit confused. How can we assume SPIR-V for -fsycl. What about -fsycl -fsycl-targets=nvptx?

mdtoguchi replied:
`-fsycl` by itself has always implied SPIR-V.  This initial support is for JIT only, which means not introducing `-fsycl-targets` yet.

bader replied:
I would phrase this differently. `-fsycl` implies that device code is compiled for some target. The target for device code is set by `--offload-arch` (i.e. `-fsycl` implies `--offload-arch`). DPC++ uses SPIR-V by default. It should be reasonable default for upstream too, unless community sets some other default value for `--offload-arch` already.

asudarsa commented:
Minor nits. Pending review from someone in the driver team and some updates requested, I think this PR is ready to go. Thanks Mike.

mdtoguchi committed:
Update a few comments

mdtoguchi commented:
Thanks for the reviews and comments.  I will be prepping a new PR for llvm-project/main based on these changes.
